File: src/main/java/com/github/ramonwirsch/fopRenderer/FopRendererPlugin.java
package com.github.ramonwirsch.fopRenderer;

import com.github.ramonwirsch.fopRenderer.util.CollectionUtil;
import org.codehaus.groovy.runtime.StringGroovyMethods;
import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.tasks.TaskContainer;

/**
 * Created by ramon on 13.11.2015.
 */
public class FopRendererPlugin implements Plugin<Project> {

	static boolean isOffline(Project project) {
		if (project.getGradle().getStartParameter().isOffline()) {
			return true;
		} else if (project.getProperties().containsKey("offlineSchemas")) {
			Object offlineSchemas = project.getProperties().get("offlineSchemas");
			if (offlineSchemas == null)
				return false;

			return Boolean.valueOf(offlineSchemas.toString());
		} else {
			return false;
		}
	}

	@Override
	public void apply(Project project) {
		project.apply(CollectionUtil.immutableMap("plugin", "base"));

		NamedDomainObjectContainer<SchemaConfigExtension> schemas = project.container(SchemaConfigExtension.class, name -> new SchemaConfigExtension(name, project));

		NamedDomainObjectContainer<RenderConfigExtension> render = project.container(RenderConfigExtension.class);

		FopRenderExtension fopRenderer = project.getExtensions().create("fopRenderer", FopRenderExtension.class, schemas, render);

		TaskContainer tasks = project.getTasks();

		Task validateAllTask = tasks.create("validate", it -> {
			it.setGroup("verification");
			it.setDescription("Validate all XML files");
		});

		tasks.findByName("check").dependsOn(validateAllTask);

		project.afterEvaluate(it -> {
			fopRenderer.getSchemas().forEach(schemaConfig -> {
				Task currentValidationTask = tasks.create("validate" + StringGroovyMethods.capitalize(schemaConfig.getName()), XMLValidatorTask.class, task -> {
					task.setSchemaConfig(schemaConfig);
				});

				validateAllTask.dependsOn(currentValidationTask);
			});

			Task buildTask = tasks.findByName("build");

			fopRenderer.getRender().forEach(renderConfig -> {

				XSLTTransformTask currentTransformTask = tasks.create("transform" + StringGroovyMethods.capitalize(renderConfig.getName()), XSLTTransformTask.class, task -> {
					task.setRenderConfig(renderConfig);
					task.dependsOn(validateAllTask);
				});

				FopRenderTask currentRenderTask = tasks.create("render" + StringGroovyMethods.capitalize(renderConfig.getName()), FopRenderTask.class, task -> {
					task.setRenderConfig(renderConfig);
					task.setInput(currentTransformTask.getOutputFile());
					task.dependsOn(currentTransformTask);
				});

				buildTask.dependsOn(currentRenderTask);
			});
		});
	}
}

File: src/main/kotlin/com/github/ramonwirsch/fopRenderer/FopRendererPlugin.kt
package com.github.ramonwirsch.fopRenderer

import com.github.ramonwirsch.fopRenderer.validation.XMLValidatorTask
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.BasePlugin
import org.gradle.api.tasks.TaskProvider

open class FopRendererPlugin : Plugin<Project> {

	override fun apply(project: Project) {
		project.pluginManager.apply(BasePlugin::class.java)

		val schemas = project.container(SchemaConfigExtension::class.java) { name -> SchemaConfigExtension(name, project) }

		val render = project.container(RenderConfigExtension::class.java) { name -> RenderConfigExtension(name, project) }

		val fopRenderer = project.extensions.create("fopRenderer", FopRenderExtension::class.java, schemas, render)

		val tasks = project.tasks

		val allValidationTasks = mutableListOf<TaskProvider<XMLValidatorTask>>()

		fopRenderer.schemas.all {
			val schemaConfig = this
			val currentValidationTask = tasks.register("validate" + schemaConfig.name.capitalize(), XMLValidatorTask::class.java) { this.schemaConfig = schemaConfig }

			allValidationTasks += currentValidationTask
		}

		val validateAllTask = tasks.register("validate") {
			group = "verification"
			description = "Validate all XML files"

			dependsOn(allValidationTasks)
		}

		tasks.named("check").configure {
			dependsOn(validateAllTask)
		}

		val allRenderTasks = mutableListOf<TaskProvider<FopRenderTask>>()

		fopRenderer.render.all {
			val renderConfig = this

			val currentTransformTask = tasks.register("transform" + renderConfig.name.capitalize(), XSLTTransformTask::class.java) {
				this.renderConfig = renderConfig

				val outputFileName = renderConfig.name

				outputFileProperty.set(project.layout.buildDirectory.file("fo/$outputFileName.fo"))

				if (renderConfig.isRequiresValidation) {
					dependsOn(validateAllTask)
				}
			}

			renderConfig.transformTask = currentTransformTask

			val currentRenderTask = tasks.register("render" + renderConfig.name.capitalize(), FopRenderTask::class.java) {
				this.renderConfig = renderConfig
				inputFileProperty.set(currentTransformTask.get().outputFileProperty)
			}

			renderConfig.renderTask = currentRenderTask

			allRenderTasks += currentRenderTask
		}

		tasks.named("build").configure {
			dependsOn(allRenderTasks)
		}
	}

	companion object {

		fun isOffline(project: Project): Boolean {
			if (project.gradle.startParameter.isOffline) {
				return true
			} else if (project.properties.containsKey("offlineSchemas")) {
				val offlineSchemas = project.properties["offlineSchemas"] ?: return false

				return java.lang.Boolean.valueOf(offlineSchemas.toString())!!
			} else {
				return false
			}
		}
	}
}